bin/dada2_preprocessing.R

```diff
@@ -0,0 +1,72 @@
+# process to each demultiplexed amplicon file.
+#
+# The primary goal is to improve the quality of the reads by removing those that 
+# are likely erroneous. This is achieved by setting various thresholds for expected
+# number of errors and length. 
+#
+# The final output is a "filtered" directory for each sample, containing the 
+# filtered demultiplexed amplicons. There will also be a single metadata file 
+# that provides a summary of the number of reads that were input and the number 
+# that remain after filtering.
+################################################################################
+
+library(argparse)
+library(dada2)
+
+# Argument Parsing and Setup
+parser <- ArgumentParser(description='Filter Errors')
+parser$add_argument('--trimmed-path', nargs='+', type="character", required=TRUE, help="Path to trimmed fastq files")
+parser$add_argument('--maxEE', type="numeric", required=TRUE, help="Maximum expected error")
+parser$add_argument('--minLen', type="numeric", required=TRUE, help="Minimum length")
+parser$add_argument('--truncQ', type="numeric", required=TRUE, help="Truncate quality")
+parser$add_argument('--maxN', type="numeric", required=TRUE, help="Maximum N")
+parser$add_argument('--rm-phix', type="logical", required=TRUE, help="Remove PhiX")
+parser$add_argument('--compress', type="logical", required=TRUE, help="Compress")
+parser$add_argument('--ncores', type="numeric", required=TRUE, help="Number of threads to use")
+parser$add_argument('--trimRight', type="numeric", required=TRUE, help="Trim Right")
+parser$add_argument('--trimLeft', type="numeric", required=TRUE, help="Trim Left")
+parser$add_argument('--matchIDs', type="logical", required=TRUE, help="Match IDs")
+parser$add_argument('--verbose', action="store_true", help="Verbose")
+
+args <- parser$parse_args()
+print(args)
+
+# Extract directory names to create corresponding output directories
+dir_names <- basename(args$trimmed_path)
+output_paths <- sprintf("filtered_%s", dir_names)
+
+# Create output directories
+lapply(output_paths, function(x) dir.create(x, recursive=TRUE, showWarnings=FALSE))
+
+# List Files and create filter paths in a vectorized manner
+fnFs_list <- lapply(args$trimmed_path, function(tp) sort(list.files(path=tp, pattern="_R1.fastq.gz", recursive=TRUE, full.names = TRUE)))
+fnRs_list <- lapply(args$trimmed_path, function(tp) sort(list.files(path=tp, pattern="_R2.fastq.gz", recursive=TRUE, full.names = TRUE)))
+
+sample_names_list <- lapply(fnFs_list, function(fnFs) sapply(strsplit(basename(fnFs), "_R1"), `[`, 1))
+
+filtFs_list <- mapply(function(op, sn) paste0(op, "/", sn, "_F_filt.fastq.gz"), output_paths, sample_names_list, SIMPLIFY = FALSE)
+filtRs_list <- mapply(function(op, sn) paste0(op, "/", sn, "_R_filt.fastq.gz"), output_paths, sample_names_list, SIMPLIFY = FALSE)
+
+# Perform Filtering and save filter metadata
+filter_metadata_list <- mapply(function(fnFs, fnRs, filtFs, filtRs) {
```

Comment:
not really a memory thing, i thought it may be useful to write this in a way so the script could accept data from multiple samples or a single sample, depending on the use case.  i don't really do that in the other scripts, i just assume it's data for a single sample, although they could be modified so that data for multiple samples could be passed in

Reply:
Oh I see, so this pattern makes it so it can run on one file or a directory? If it doesn't cause a perf hit that makes sense to me

Reply:
mostly just more than one directory. i thought writing it this way would allow some flexibility. there shouldn't be any added performance problems from this, just an expectation that runtime will grow linearly if you decide to pass in more than one sample directory